Updated EMS docs (#6855)
expo/expo: packages/expo-module-scripts/README.md

**This is the package that installs Babel CLI, TypeScript, Jest, and other common development dependencies.** Update the dependencies in this package when changing them for the Expo repository.

- [Getting Started](#getting-started)
- [Setup](#setup)
  - [🤡 Jest](#-jest)
  - [📝 LICENSE](#-license)

>   - [Side-Effects](#side-effects)

Reviewer: ```suggestion
  - [Side Effects](#side-effects)
```

  - [Entry Point and Types](#entry-point-and-types)
  - [🔗 NPM Linking](#-npm-linking)
- [⌘ Commands](#-commands)
  - [configure](#configure)
  - [typecheck](#typecheck)
  - [build](#build)
  - [test](#test)
  - [lint](#lint)
  - [clean](#clean)
- [Lifecycle Commands](#lifecycle-commands)
  - [prepare (npm lifecycle)](#prepare--npm-lifecycle-)
  - [prepublishOnly (npm lifecycle)](#prepublishonly--npm-lifecycle-)
- [Unified dependencies](#unified-dependencies)

## Getting Started

```sh
yarn add -D expo-module-scripts

# or

npm install --save-dev expo-module-scripts
```

## Setup

Add the following scripts to your `package.json` and run `yarn`

```json5
{ 
  "scripts": {
      "build": "expo-module build",
      "clean": "expo-module clean",
      "test": "expo-module test",
      "prepare": "expo-module prepare",
      "prepublishOnly": "expo-module prepublishOnly",
      "expo-module": "expo-module"
  },
}
```

> Running `yarn` will now run the `prepare` script which generates any missing files:

Reviewer: ```suggestion
Running `yarn` will now run the `prepare` script, which generates any missing files:
```

- [`.eslintrc.js`](./templates/.eslintrc.js) ([docs](https://eslint.org/docs/user-guide/configuring)) this extends [`eslint-config-universe`](https://github.com/expo/expo/tree/master/packages/eslint-config-universe).

>   - Optionally you can customize prettier too: [.prettierrc guide](https://github.com/expo/expo/tree/master/packages/eslint-config-universe#customizing-prettier).

Reviewer: ```suggestion
  - Optionally you can customize Prettier too: [.prettierrc guide](https://github.com/expo/expo/tree/master/packages/eslint-config-universe#customizing-prettier).
```

- [`.npmignore`](./templates/.npmignore) ([docs](https://docs.npmjs.com/misc/developers)) currently only ignores the `babel.config.js` in your module. You might also want to also add tests and docs.
  - Expo modules use `.npmignore` **instead of** the `files` field in the `package.json`.

>   - (Pro-Tip) Test which files get packaged by running `npm pack`, if you see files that aren't crucial to running the module you should add them to the `.npmignore`.

Reviewer: ```suggestion
  - (Pro Tip) Test which files get packaged by running `npm pack`. If you see files that aren't crucial to running the module, you should add them to `.npmignore`.
```

- [`README.md`](./templates/README.md) A default template for Unimodule installation.
  - Project docs should try to have relevant emojis in headers because OSS is fun.
  - Use [badges](https://github.com/expo/expo#-badges)
  - Try and incorporate a table of contents (TOC).
- [`tsconfig.json`](./templates/tsconfig.json) ([docs](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html)) extends [`tsconfig.base.json`](./tsconfig.base.json) this is important for ensuring all Unimodules use the same version of TypeScript.

You should also add the following fields to your `package.json`:

### 🤡 Jest

> The jest preset extends [`jest-expo`](https://github.com/expo/expo/tree/master/packages/jest-expo) or [`jest-expo-enzyme`](https://github.com/expo/expo/tree/master/packages/jest-expo-enzyme) and adds proper TypeScript support and Types to the presets.

Reviewer: ```suggestion
The Jest preset extends [`jest-expo`](https://github.com/expo/expo/tree/master/packages/jest-expo) or [`jest-expo-enzyme`](https://github.com/expo/expo/tree/master/packages/jest-expo-enzyme) and adds proper TypeScript support and type declarations to the presets.
```

> **For unit testing API based modules:**

Reviewer: ```suggestion
**For unit testing API-based modules:**
```

```json5
{
  "jest": {
    "preset": "expo-module-scripts/universal"
  },
}
```

> **For unit testing Component based modules:**

Reviewer: ```suggestion
**For unit testing component-based modules:**
```

```json5
{
  "jest": {
    "preset": "expo-module-scripts/enzyme"
  },
}
```

### 📝 LICENSE

> This makes it easier for other members of the community to work with your package. Expo usually has **MIT** licenses.

Reviewer: ```suggestion
This makes it easier for other members of the community to work with your package. Expo usually has the **MIT** license.
```

```json5
{
  "license": "MIT",
}
```

> ### Side-Effects

Reviewer: ```suggestion
### Side Effects
```

> The [`@expo/webpack-config`](https://www.npmjs.com/package/@expo/webpack-config) is optimized for tree-shaking, you should always make sure to list whatever files in your module have side-effects. In Expo modules we use the `.fx.*` extension on these files (this makes it easier to target them with `sideEffects`).

Reviewer: ```suggestion
The [`@expo/webpack-config`](https://www.npmjs.com/package/@expo/webpack-config) is optimized for tree-shaking, you should always make sure to list whatever files in your module have side effects. In Expo modules we use the `.fx.*` extension on these files (this makes it easier to target them with `sideEffects`).
```

> [**Learn more about side-effects**](https://webpack.js.org/guides/tree-shaking/)

Reviewer: ```suggestion
[**Learn more about side effects**](https://webpack.js.org/guides/tree-shaking/)
```
